### PR TITLE
Support types like `owned class`

### LINKED ISSUE #13088: Specifying arguments with class type and any nilability?
Should one be able to create a generic function argument that accepts both `MyClass` and `MyClass?`? Or `owned` and `owned?` ?

Currently, one can create functions generic on `owned`. When `owned` might be nilable or non-nilable, following the general strategy in #12614, `owned` will become "non-nilable owned" just as `MyClass` means "non-nilable MyClass".

I'm not sure if such functions will be generally important, but they are needed in the implementation of `owned` and `shared` and any user-provided memory management types.

### PR DESCRIPTION
**Pull request by mppf**
Closes https://github.com/Cray/chapel-private/issues/411
Resolves #13088

This PR implements support for type declarations like `owned class`. Here `class` takes the place of a specific class and so means the same thing with regard to nilability. This design is discussed in #13088.

* factored `extern _` parse rules into one rule. Now that `record` can be a type expression, `extern record R` and `extern record proc f()`are ambiguous. The latter case comes up with the extern-proc-renaming where an arbitrary `expr` is allowed (see #10298)
* `extern class` is now a parse error (it has been an error since  #11219). The compiler no longer has specific support for `extern class` and it has been deprecated for 2 releases. Therefore, I removed the `test/deprecated/extern-classes.chpl` test. I also removed some compiler code that obviously only applied to extern classes.

Reviewed by @vasslitvinov - thanks!

- [x] full local futures testing

### CODE REVIEW

**Comment by mppf** [comment]
[Error matching program output for classes/delete-free/borrow-resolution]
[Error matching program output for classes/delete-free/type-casting]
[Error matching program output for classes/initializers/generics/declarations/no-arg-type5]
[Error matching program output for classes/nilability/nilability-coercions-owned-shared (compopts: 1)]
[Error matching program output for classes/nilability/nilability-coercions-owned-shared (compopts: 2)]
[Error matching program output for classes/nilability/nilability-coercions]

File: test/classes/nilability/nilability-upcasting.chpl [hunk 1]
```diff
@@ -128,6 +128,8 @@ module test {
         writeln(x.type:string, " ", x);
         var y = cb:borrowed;
         writeln(y.type:string, " ", y);
+        var z = cb:borrowed class;
+        writeln(y.type:string, " ", y);
```
Comment (vasslitvinov) [inline test/classes/nilability/nilability-upcasting.chpl]: Please `z` and not `y` :)

I checked the other cases you added in this test - they look correct in this regard.

File: test/deprecated/extern-classes.chpl [hunk 2]
```diff
@@ -1,8 +0,0 @@
-
-extern class C {
-  var x : int;
-}
-
-var c : C;
-writeln(c == nil);
-
```
Comment (vasslitvinov) [inline test/deprecated/extern-classes.chpl]: Please explain in the PR message why you deleted this test.

File: test/extern/bradc/classes/extern_subclasses.good [hunk 3]
```diff
@@ -1,2 +1 @@
-extern_subclasses.chpl:5: error: External types do not currently support inheritance
-extern_subclasses.chpl:13: error: External types do not currently support inheritance
+extern_subclasses.chpl:5: syntax error: near 'C2'
```
Comment (vasslitvinov) [inline test/extern/bradc/classes/extern_subclasses.good]: Ugh it was nicer before.
Not a show stopper though. Would be nice to add an "error message" future.

File: test/chpldoc/nodoc/privateClasses.doc.bad [hunk 4]
```diff
@@ -1,5 +1 @@
-privateClasses.doc.chpl:1: syntax error: near 'class'
-privateClasses.doc.chpl:4: syntax error: near '}'
-privateClasses.doc.chpl:10: syntax error: near '}'
-privateClasses.doc.chpl:22: syntax error: near '}'
 cat: docs/source/modules/privateClasses.rst: No such file or directory
```
[the remark this answers is not in the document]
Comment (mppf) [reply test/chpldoc/nodoc/privateClasses.doc.bad]: The compiler no longer issues an error for `private class X { }` *when running chpldoc*.  It does issue the error when compiling. (Why this is - is unclear to me - but I don't think it's a big issue. Also, we have other tests that check for this error when actually compiling).

```
$ chpl test/chpldoc/nodoc/privateClasses.doc.chpl 
test/chpldoc/nodoc/privateClasses.doc.chpl:1: error: Can't apply private to types yet
test/chpldoc/nodoc/privateClasses.doc.chpl:7: error: Can't apply private to types yet
test/chpldoc/nodoc/privateClasses.doc.chpl:15: error: Can't apply private to the fields or methods of a class or record yet
test/chpldoc/nodoc/privateClasses.doc.chpl:18: error: Can't apply private to the fields or methods of a class or record yet
```

File: test/extern/bradc/classes/extern_subclasses.good (hunk 3 above)
Comment (mppf) [reply test/extern/bradc/classes/extern_subclasses.good]: `extern class` has been deprecated more than 1 release. If `extern class` is not part of the language, I don't see any particular reason why we need to expend energy making a better error message for it.

File: test/classes/nilability/nilability-upcasting.chpl (hunk 1 above)
Comment (mppf) [reply test/classes/nilability/nilability-upcasting.chpl]: Thanks for catching that!

**Pull request merged**